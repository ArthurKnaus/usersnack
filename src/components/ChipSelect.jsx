import PropTypes from 'prop-types';
import { withStyles } from 'material-ui/styles';

import { Select, Chip, MenuItem, FormControl, InputLabel, Input } from 'material-ui';

const ITEM_HEIGHT = 48;
const ITEM_PADDING_TOP = 8;

const styles = theme => ({
  formControl: {
    marginTop: theme.spacing.unit,
    width: '100%',
  },
  chips: {
    display: 'flex',
    flexWrap: 'wrap',
  },
  chip: {
    margin: theme.spacing.unit / 4,
  },
  activeItem: {
    fontWeight: theme.typography.fontWeightMedium,
  },
});

const ChipSelect = props => (
  <FormControl className={props.classes.formControl}>
    <InputLabel htmlFor={props.id}>
      {props.label}
    </InputLabel>
    <Select
      multiple
      value={props.value}
      input={<Input id={props.id} />}
      onChange={event => props.onChange(event.target.value)}
      renderValue={selected => (
        <div className={props.classes.chips}>
          {selected.map(item => (
            <Chip
              key={item}
              label={item}
              onDelete={() => props.onChange(props.value.filter(i => i !== item))}
              className={props.classes.chip}
            />
          ))}
        </div>
      )}
      MenuProps={{
        PaperProps: {
          style: {
            maxHeight: (ITEM_HEIGHT * 6.5) + ITEM_PADDING_TOP,
            width: 250,
          },
        },
      }}
    >
      {props.items.map(item => (
        <MenuItem
          key={item}
          value={item}
          className={props.value.indexOf(item) > -1
            ? props.classes.activeItem
            : undefined
          }
        >
          {item}
        </MenuItem>
      ))}
    </Select>
  </FormControl>
);

ChipSelect.propTypes = {
  id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  items: PropTypes.arrayOf(PropTypes.string).isRequired,
  value: PropTypes.arrayOf(PropTypes.string).isRequired,
  onChange: PropTypes.func.isRequired,
  classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(ChipSelect);
