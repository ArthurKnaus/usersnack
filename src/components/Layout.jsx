import { Grid, AppBar, Toolbar, Typography } from 'material-ui';
import { withStyles } from 'material-ui/styles';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

const styles = theme => ({
  contentWrapper: {
    paddingTop: (theme.spacing.unit * 6) + 64,
    [theme.breakpoints.down('sm')]: {
      paddingTop: theme.spacing.unit + 56,
    },
    [theme.breakpoints.up('xl')]: {
      paddingTop: (theme.spacing.unit * 8) + 64,
    },
  },
});

const Layout = props => (
  <div>
    <AppBar>
      <Toolbar>
        <Typography variant="title" color="inherit">
          { props.title }
        </Typography>
      </Toolbar>
    </AppBar>
    <Grid container justify="center" className={props.classes.contentWrapper}>
      <Grid item xs={12} md={10} lg={8} xl={8}>
        { props.children }
      </Grid>
    </Grid>
  </div>
);

Layout.propTypes = {
  children: PropTypes.node.isRequired,
  title: PropTypes.string.isRequired,
  classes: PropTypes.shape({
    contentWrapper: PropTypes.string,
  }).isRequired,
};

const mapStateToProps = state => ({
  title: state.app.title,
});

export default withStyles(styles)(connect(mapStateToProps)(Layout));

// export without higher-order-components for testing
const PlainLayout = Layout;
export {
  PlainLayout,
};
