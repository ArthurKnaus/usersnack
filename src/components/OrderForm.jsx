import PropTypes from 'prop-types';

import { TextField, Button } from 'material-ui';
import SendIcon from 'material-ui-icons/Send';

import { withStyles } from 'material-ui/styles';

const styles = () => ({
  container: {
    display: 'flex',
    flexWrap: 'wrap',
  },
  input: {
    marginRight: 8,
    flex: '1 0 auto',
  },
  button: {
    height: 40,
    marginTop: 24,
  },
  icon: {
    marginLeft: 8,
  },
});

const onEventValue = (key, f) => event => f(key, event.target.value);

const OrderForm = props => (
  <div className={props.classes.container}>
    <TextField
      className={props.classes.input}
      onChange={onEventValue('name', props.onValueChange)}
      required
      id="name"
      label="Name"
      value={props.values.name}
      error={props.errors.name}
      margin="normal"
    />
    <TextField
      className={props.classes.input}
      onChange={onEventValue('address', props.onValueChange)}
      required
      id="address"
      label="Address"
      value={props.values.address}
      error={props.errors.address}
      margin="normal"
    />
    <Button
      className={props.classes.button}
      onClick={props.onSubmit}
      variant="raised"
      color="primary"
    >
      Order
      <SendIcon className={props.classes.icon} />
    </Button>
  </div>
);

OrderForm.propTypes = {
  onValueChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  values: PropTypes.shape({
    name: PropTypes.string,
    address: PropTypes.string,
  }).isRequired,
  errors: PropTypes.shape({
    name: PropTypes.bool,
    address: PropTypes.bool,
  }).isRequired,
  classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(OrderForm);
