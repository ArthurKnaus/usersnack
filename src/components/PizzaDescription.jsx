import { IconButton, CardContent, Typography, CardMedia, CardHeader } from 'material-ui';
import NavigateBeforeIcon from 'material-ui-icons/NavigateBefore';
import { withStyles } from 'material-ui/styles';
import PropTypes from 'prop-types';

const formatter = new Intl.NumberFormat('en-US', { style: 'currency', currency: 'EUR' });

const styles = () => ({
  cardHeader: {
    paddingTop: 0,
    paddingLeft: 24,
  },
  backButton: {
    position: 'absolute',
    height: 32,
    width: 32,
    top: 16,
    left: 8,
  },
  descriptionWrapper: {
    position: 'relative',
    display: 'flex',
  },
  description: {
    flex: '1 1 auto',
  },
  image: {
    minWidth: 167,
    minHeight: 167,
    maxHeight: 167,
  },
});

const PizzaDescription = (props) => {
  const { classes, pizza, onBack } = props;
  return (
    <CardContent className={classes.descriptionWrapper}>
      <div className={classes.description}>
        <IconButton
          className={classes.backButton}
          color="primary"
          onClick={onBack}
        >
          <NavigateBeforeIcon />
        </IconButton>
        <CardHeader
          className={classes.cardHeader}
          title={pizza.name}
          subheader={`Base price: ${formatter.format(pizza.price)}`}
        />
        <Typography variant="subheading" color="textSecondary">
          Ingredients:
        </Typography>
        <Typography variant="body2">
          {pizza.ingredients.join(', ')}
        </Typography>
      </div>
      <CardMedia
        className={classes.image}
        image={`/static/${pizza.img}`}
        title={pizza.name}
      />
    </CardContent>
  );
};

PizzaDescription.propTypes = {
  pizza: PropTypes.shape({
    name: PropTypes.string,
    ingredients: PropTypes.arrayOf(PropTypes.string),
  }).isRequired,
  onBack: PropTypes.func.isRequired,
  classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(PizzaDescription);
