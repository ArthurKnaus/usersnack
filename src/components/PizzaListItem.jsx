import { Card, CardContent, Typography, CardMedia, ButtonBase } from 'material-ui';
import { withStyles } from 'material-ui/styles';
import PropTypes from 'prop-types';

const formatter = new Intl.NumberFormat('en-US', { style: 'currency', currency: 'EUR' });

const styles = () => ({
  card: {
    display: 'flex',
    marginBottom: 16,
    minHeight: 151,
  },
  button: {
    width: '100%',
    textAlign: 'inherit',
    alignItems: 'normal',
    justifyContent: 'normal',
  },
  details: {
    display: 'flex',
    flex: '1 1 auto',
    flexDirection: 'column',
  },
  content: {
    flex: '1 0 auto',
    display: 'flex',
    flexDirection: 'column',
  },
  ingredients: {
    flex: '1 0 auto',
  },
  image: {
    minWidth: 151,
    minHeight: 151,
    borderTopRightRadius: 2,
    borderBottomRightRadius: 2,
  },
});

const PizzaListItem = (props) => {
  const { classes, pizza, onClick } = props;
  return (
    <Card className={classes.card}>
      <ButtonBase
        className={classes.button}
        onClick={onClick}
      >
        <div className={classes.details}>
          <CardContent className={classes.content}>
            <Typography variant="headline">
              {pizza.name}
            </Typography>
            <Typography className={classes.ingredients} variant="subheading" color="textSecondary">
              {pizza.ingredients.join(', ')}
            </Typography>
            <Typography variant="subheading" color="textSecondary">
              {formatter.format(pizza.price)}
            </Typography>
          </CardContent>
        </div>
        <CardMedia
          className={classes.image}
          image={`static/${pizza.img}`}
          title={pizza.name}
        />
      </ButtonBase>
    </Card>
  );
};

PizzaListItem.propTypes = {
  pizza: PropTypes.shape({
    name: PropTypes.string,
    ingredients: PropTypes.arrayOf(PropTypes.string),
  }).isRequired,
  onClick: PropTypes.func.isRequired,
  classes: PropTypes.shape({
    card: PropTypes.string,
    details: PropTypes.string,
    content: PropTypes.string,
  }).isRequired,
};

export default withStyles(styles)(PizzaListItem);
