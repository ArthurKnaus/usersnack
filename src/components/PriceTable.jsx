import PropTypes from 'prop-types';

import Table, { TableHead, TableBody, TableRow, TableCell, TableFooter } from 'material-ui/Table';

const formatter = new Intl.NumberFormat('en-US', { style: 'currency', currency: 'EUR' });

const sumExtras = (basePrice, extras) => {
  let sum = basePrice;

  if (extras.length > 0) {
    sum += extras.map(e => e.price)
      .reduce((a, b) => a + b);
  }

  // Needed because of floating point precision problem
  return formatter.format(sum);
};

const PriceTable = props => (
  <Table>
    <TableHead>
      <TableRow>
        <TableCell>Item</TableCell>
        <TableCell numeric>Price</TableCell>
      </TableRow>
    </TableHead>
    <TableBody>
      <TableRow>
        <TableCell>Base price</TableCell>
        <TableCell numeric>{formatter.format(props.basePrice)}</TableCell>
      </TableRow>
      {props.extras.map(extra => (
        <TableRow key={extra.name}>
          <TableCell>{extra.name}</TableCell>
          <TableCell numeric>{formatter.format(extra.price)}</TableCell>
        </TableRow>
      ))}
    </TableBody>
    <TableFooter>
      <TableRow>
        <TableCell>Total</TableCell>
        <TableCell numeric>{sumExtras(props.basePrice, props.extras)}</TableCell>
      </TableRow>
    </TableFooter>
  </Table>
);

PriceTable.propTypes = {
  extras: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string,
    price: PropTypes.number,
  })).isRequired,
  basePrice: PropTypes.number.isRequired,
};

export default PriceTable;
