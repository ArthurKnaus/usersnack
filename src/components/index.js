import Layout from './Layout';
import PizzaListItem from './PizzaListItem';
import PizzaDescription from './PizzaDescription';
import ChipSelect from './ChipSelect';
import PriceTable from './PriceTable';
import OrderForm from './OrderForm';

export {
  Layout,
  PizzaListItem,
  PizzaDescription,
  ChipSelect,
  PriceTable,
  OrderForm,
};
