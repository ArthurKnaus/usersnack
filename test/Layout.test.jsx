import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import { Layout } from '../src/components/Layout';

Enzyme.configure({ adapter: new Adapter() });

const title = 'testTitle';
const child = <div id="child" />;
const props = {
  title,
  classes: {},
};

describe('PizzaForm', () => {
  it('should match snapshot', () => {
    const component = renderer.create(<Layout {...props}>{child}</Layout>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render with child and text', () => {
    const wrapper = shallow(<Layout {...props}>{child}</Layout>);
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.contains(child)).toBe(true);
    expect(wrapper.contains(title)).toBe(true);
  });
});
