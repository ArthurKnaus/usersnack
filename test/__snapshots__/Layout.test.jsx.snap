// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PizzaForm should match snapshot 1`] = `
<div>
  <header
    className="MuiPaper-root-9 MuiPaper-shadow4-15 MuiAppBar-root-1 MuiAppBar-positionFixed-2 MuiAppBar-colorPrimary-7 mui-fixed"
  >
    <div
      className="MuiToolbar-root-36 MuiToolbar-gutters-37"
    >
      <h2
        className="MuiTypography-root-38 MuiTypography-title-44 MuiTypography-colorInherit-57"
      >
        testTitle
      </h2>
    </div>
  </header>
  <div
    className="MuiGrid-typeContainer-62 MuiGrid-spacing-xs-16-84 MuiGrid-justify-xs-center-79"
  >
    <div
      className="MuiGrid-typeItem-63 MuiGrid-grid-xs-12-99 MuiGrid-grid-md-10-123 MuiGrid-grid-lg-8-134 MuiGrid-grid-xl-8-147"
    >
      <div
        id="child"
      />
    </div>
  </div>
</div>
`;
